crates/typst-library/src/foundations/array.rs

```diff
@@ -841,31 +841,63 @@ impl Array {
         /// determine the keys to sort by.
         #[named]
         key: Option<Func>,
+        /// If given, uses this function to compare elements in the array.
+        ///
+        /// This function should return an integer, whose sign is used to
+        /// determine the relative order of two given elements: Negative if the
+        /// first element is smaller, positive if the second element is smaller.
+        /// If `{0}` is returned, the order of the elements is not modified.
+        ///
+        /// When used together with `key`, `compare` will be passed the keys
+        /// instead of the elements.
+        ///
+        /// ```example
+        /// #(
+        ///   "sorted",
+        ///   "by",
+        ///   "decreasing",
+        ///   "length",
+        /// ).sorted(
+        ///   key: s => s.len(),
+        ///   compare: (x, y) => y - x,
+        /// )
+        /// ```
+        #[named]
+        compare: Option<Func>,
     ) -> SourceResult<Array> {
         let mut result = Ok(());
         let mut vec = self.0;
-        let mut key_of = |x: Value| match &key {
-            // NOTE: We are relying on `comemo`'s memoization of function
-            // evaluation to not excessively reevaluate the `key`.
-            Some(f) => f.call(engine, context, [x]),
-            None => Ok(x),
+        let mut compare = |x: Value, y: Value| {
+            let mut key_of = |x: Value| match &key {
+                // NOTE: We are relying on `comemo`'s memoization of function
+                // evaluation to not excessively reevaluate the `key`.
+                Some(f) => f.call(engine, context, [x]),
+                None => Ok(x),
+            };
+            let x = key_of(x)?;
+            let y = key_of(y)?;
+            match &compare {
+                Some(f) => Ok(match f.call(engine, context, [x, y])? {
+                    Value::Int(x) => x.cmp(&0),
+                    x => bail!(
+                        span,
+                        "expected integer from `compare` function; got {}",
+                        x.repr()
```

Comment:
```suggestion
                        x.ty()
```
typically, we show the type rather than the value.

could you also add a test for this case?